### Try Model: Tensor Flow Neural Network

In [1]:
# setup absolute path to location of package Starts and config-file 
from inspect import getsourcefile
import os.path as path, sys
current_dir = path.dirname(path.abspath(getsourcefile(lambda:0)))
sys.path.insert(0, current_dir[:current_dir.rfind(path.sep)])

# import package Starts
from Starts.startmodskl import *
from Starts.startmodtf import *

local_kwargs {'data_path_1': './data/train.csv', 'data_path_2': './data/titanic_no_nan.csv', 'data_path_3': './data/test.csv', 'drop_obj_col': False, 'nan_drop_col': False, 'nan_drop_row': False, 'nan_zero': False, 'nan_mean': True, 'nan_mean_neighbors': False}


Using TensorFlow backend.


In [2]:
nonan_data.head()

Survived  Pclass  Sex        Age  Ticket    Fare  Title  FamilySize
0         1       2    0  29.000000     553  10.500      6           0
1         0       2    1  23.000000     119  13.000      5           0
2         0       3    1  28.000000     306   9.500      5           0
3         0       3    1  20.000000     500   9.225      5           0
4         0       1    1  29.699118      13   0.000      5           0

In [3]:
stf = StartModTF(nonan_data)
# Try method Keras Tensorflow to predict the Survivors (column 'Survived')
model, y_true, y_pred = stf.keras_sequential(nonan_data, 'Survived', hidden_layers=2)

Epoch 1/1
712/712 [==============================] - 1s 1ms/step - loss: 0.6818 - acc: 0.6208


In [4]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])

# Compare the predicted value with true_value
len(y_true), len(y_pred)

(179, 179)

In [5]:
confusion_matrix(y_true, y_pred)

array([[105,   2],
       [ 72,   0]])

### Result: 
#### the correct predicted result 105
#### the wrong predicted result 74

In [6]:
# update parameters
new_param={'hidden_units':[10,10,10], 'optimizer':'Adam', 'activation_fn':'sigmoid', 'learning_rate': 0.0001,
           'steps':2000, 'batch_size':10, 'num_epochs':10, 'feature_scl':True}
stf.update_parameters=new_param
model, y_true, y_pred = stf.keras_sequential(nonan_data, 'Survived', hidden_layers=2)

Epoch 1/10
712/712 [==============================] - 0s 428us/step - loss: 0.6794 - acc: 0.6208
Epoch 2/10
712/712 [==============================] - 0s 140us/step - loss: 0.6408 - acc: 0.6208
Epoch 3/10
712/712 [==============================] - 0s 144us/step - loss: 0.6335 - acc: 0.6208
Epoch 4/10
712/712 [==============================] - 0s 137us/step - loss: 0.6258 - acc: 0.6208
Epoch 5/10
712/712 [==============================] - 0s 141us/step - loss: 0.6192 - acc: 0.6208
Epoch 6/10
712/712 [==============================] - 0s 167us/step - loss: 0.6124 - acc: 0.6433
Epoch 7/10
712/712 [==============================] - 0s 178us/step - loss: 0.6042 - acc: 0.7121
Epoch 8/10
712/712 [==============================] - 0s 137us/step - loss: 0.5929 - acc: 0.7177
Epoch 9/10
712/712 [==============================] - 0s 142us/step - loss: 0.5845 - acc: 0.7247
Epoch 10/10
712/712 [==============================] - 0s 157us/step - loss: 0.5735 - acc: 0.7261


In [7]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])
confusion_matrix(y_true, y_pred)

array([[93, 14],
       [37, 35]])

### Result: 
#### the correct predicted result 128
#### the wrong predicted result 51 

### Result has been changed (getting worse) without any impact from code, so data preprocessing has major impact on final result. In our case, the feature 'Ticket' was not removed during preprocess phase (NaN processing) and included into data. 

### Conclusion: unnecessary feature might be removed before building model.